packages/gatsby/src/internal-plugins/query-runner/query-queue.js

```diff
@@ -61,8 +61,22 @@ const queue = new Queue((plObj, callback) => {
     )
 }, queueOptions)
 
+// Pause running queries when new nodes are added (processing starts).
+emitter.on(`CREATE_NODE`, () => {
+  queue.pause()
```

Comment:
This could also be converted to use xstate

Reply:
This can't pause task that is already in progress, right? Just will stop any queued tasks from executing? So there is case where we still will be executing query in parallel to CREATE_NODE processing?

Reply:
Correct. Though in practice this is fine as most queries are sync and what's not sync is background jobs or fetching data from elsewhere, neither of which would be affectes by node transformations.

Reply:
Right, but markdown processing is async (and most of reports about problem this PR solves is about markdown). That's why I was asking if we should delay dispatching `CREATE_NODE` action while query are executed.

Reply:
If a query of a markdown node is in progress, it gets the markdown node syncly (this should be a word) at the start so even if a new file node immediately causes the markdown remark node to be deleted, it won't affect the in-progress query.

That being said, querying will eventually be fully async so we'll need a better solution for that. I think the current PR is fine as it's simple and solves the problem. Next action would be to add tests and then ensure we have a solution when we go async e.g. delaying API/redux work work until the current queries are finished.

Reply:
Good point, I was looking in wrong spot - at the transformer level we are already shielded from this issue as node is already passed. But before that we use `run-sift` to actually get the node and this one is async as well

Reply:
I think this PR seems good as-is, just trying to see if we need to pursue further